### PR TITLE
Add support for loading CA data from base64

### PR DESCRIPTION
**Pull request by jrostand**
This allows a user to paste in base64-encoded CA certificates, similar to kubeconfig files. With this enhancement, users can freely share their Osprey config files within their organization without worrying about file paths that can change between machines.

### CODE REVIEW

**Comment by totahuanocotl** [comment]
Thanks for your PR @jrostand. I'll accept the changes, but before I can merge the PR, there are a few things that need to be changed.
- I think the use case of your PR is not entirely clear since this component is used by both the client and the server. Please decide for which component you are submitting this change, it could be one or both.
  Once that has been defined, to make it clear that the expected value is a base64 encoded string I would ask that you:
  - add a new optional flag to the serve command `apiServerCAData`,  if you want this change for the server. 
  - update the osprey client config to accept an optional `certificate-authority-data` field (just as kubeconfig does), if you want this change for the client. 
  If the user provides both values the `data` input should override the file one (please describe this in the command flag description).
- Please add tests for these scenarios.
- Update the README documentation for client/server. 
- Add a CHANGELOG entry for this change. 

Thanks again for your contribution!

**Comment by jrostand** [comment]
@totahuanocotl Fair points all. This is intended as a client change, so I'll get on to making those requested changes shortly. Thanks for the input.

**Comment by jrostand** [comment]
Changes pushed. Hopefully it's close to what you had in mind.

**Comment by jrostand** [comment]
@totahuanocotl I think my latest commit should just about cover all the changes requested. Thanks for the help on this - Go is one of my lesser-used languages, so I'm not great at designing with it, particularly in an established codebase.

File: e2e/e2e_test.go [hunk 1]
```diff
@@ -198,6 +186,40 @@ var _ = Describe("E2E", func() {
...
+				BeforeEach(func() {
+					caDataConfig, err := BuildCADataConfig(testDir, ospreys, true, "")
+					Expect(err).To(BeNil(), "Creates the osprey config")
+
+					caDataConfigFlag := "--ospreyconfig=" + caDataConfig.ConfigFile
+
+					caDataLogin = Client("user", "login", caDataConfigFlag)
+				})
+
+				It("logs in successfully", func() {
```
Comment (totahuanocotl) [inline e2e/e2e_test.go]: Since it is just one test scenario for this context, I think having a beforeEach  and a context makes it harder to read.  
We can get rid of the context, merge their names and put the contents of the before each inside this function. 
`It ("logs in using certificate-authority-data")` reads well enough. You can remove the empty line spaces in from the setup block, so that it is clear that it is a unit of work, and just keep one between that an the login.

File: e2e/e2e_test.go [hunk 2]
```diff
@@ -198,6 +186,40 @@ var _ = Describe("E2E", func() {
...
+				BeforeEach(func() {
+					caDataConfig, err := BuildCADataConfig(testDir, ospreys, true, "/road/to/nowhere")
+					Expect(err).To(BeNil(), "Creates the osprey config")
+
+					caDataConfigFlag := "--ospreyconfig=" + caDataConfig.ConfigFile
+
+					caDataLogin = Client("user", "login", caDataConfigFlag)
+				})
+
+				It("ignores CA path and logs in successfully", func() {
```
Comment (totahuanocotl) [inline e2e/e2e_test.go]: same comment as above. 
`"logs in using overriding certificate-authority with certificate-authority-data "` ?

File: common/web/client.go [hunk 3]
```diff
@@ -4,14 +4,28 @@ import (
...
 	"net"
 	"net/http"
-	"os"
 	"time"
 )
 
+// LoadTLSCert loads a PEM-encoded certificate from file and returns it as a
+// base64-encoded string.
+func LoadTLSCert(path string) (string, error) {
+	fileData, err := ioutil.ReadFile(path)
```
Comment (totahuanocotl) [inline common/web/client.go]: no need for the empty lines in this method, makes it a bit harder to read.

File: cmd/serve.go [hunk 4]
```diff
@@ -55,7 +55,18 @@ func init() {
 
 func serve(cmd *cobra.Command, args []string) {
 	var err error
-	httpClient, err := webClient.NewTLSClient(issuerCA, tlsCert)
+
```
Comment (totahuanocotl) [inline cmd/serve.go]: no need for this empty line I think

File: client/config.go [hunk 5]
```diff
@@ -76,22 +77,31 @@ func LoadConfig(path string) (*Config, error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("invalid config %s: %v", path, err)
 	}
 
-	for _, target := range config.Targets {
-		// Overwrite target CA if we have a base64 CA cert
-		if target.CertificateAuthorityData != "" {
-			target.CertificateAuthority = target.CertificateAuthorityData
+	// Set CAData from PEM-encoded cert file if no CAData
+	if config.CertificateAuthority != "" && config.CertificateAuthorityData == "" {
```
Comment (totahuanocotl) [inline client/config.go]: I think you left out the part where, if CAData is present, we should set config.CertificateAuthority to "", so that it is clear that the data is unrelated to that CA path.

File: client/config.go [hunk 6]
```diff
@@ -76,22 +77,31 @@ func LoadConfig(path string) (*Config, error) {
...
+		config.CertificateAuthorityData = certData
 	}
 
-	err = config.validate()
-	if err != nil {
-		return nil, fmt.Errorf("invalid config %s: %v", path, err)
+	for name, target := range config.Targets {
+		// Set CAData from PEM-encoded cert file if no CAData
+		if target.CertificateAuthority != "" && target.CertificateAuthorityData == "" {
+			certData, err := web.LoadTLSCert(target.CertificateAuthority)
```
Comment (totahuanocotl) [inline client/config.go]: same comment as above ragarding CAData present.

**Comment by jrostand** [comment]
@totahuanocotl latest commit should cover everything from the second review

**Comment by totahuanocotl** [comment]
@jrostand it has been released. 
Thanks!
